src/analyzing/CFG_build.ml

```diff
@@ -228,58 +238,115 @@ let rec cfg_stmt : state -> F.nodei option -> stmt -> cfg_stmt_result =
+        cfg_stmt_list try_state (Some newi) try_st
+      in
+
+      (* We do not directly connect the end of try with catchesi but whenever we
+       * encounter a Call or a Throw inside a Try, we add an arc.
+       * But if there is a path that can reach the end of try, we add an arc
+       * to elsei. *)
+      state.g |> add_arc_from_opt (finaltry, elsei);
+
+      (* In the else clause, if an exception is thrown, go to the finally
+       * clause, or if there is no finally clause, propagate the exception.
+       *)
+      let has_finally =
+        match finally_st with
+        | [] -> false
+        | _some_list_ -> true
+      in
+      let throw_destination =
+        if has_finally then Some newfakefinally else state.throw_destination
+      in
+      let else_state = { state with throw_destination } in
+      let finalelse, else_may_throw =
+        cfg_stmt_list else_state (Some elsei) else_st
+      in
+
+      state.g |> add_arc_from_opt (finalelse, newfakefinally);
+
+      (* From catches to finally in case of uncaught exceptions. *)
       state.g |> add_arc (catchesi, newfakefinally);
-      catches
-      |> List.iter (fun (_, catch_st) ->
-             let finalcatch = cfg_stmt_list state (Some catchesi) catch_st in
-             state.g |> add_arc_from_opt (finalcatch, newfakefinally));
-      let finalfinally = cfg_stmt_list state (Some newfakefinally) finally_st in
-      (* If we're inside another Try then we assume that we could propagate up
-       * some unhandled exception. Otherwise we assume that we handled everything.
-       * THINK: Alternatively, we could add an arc to the exit node. *)
-      state.g |> add_arc_opt_to_opt (finalfinally, state.try_catches_opt);
-      CfgFirstLast (newi, finalfinally)
+
+      (* Same reasoning as the else clause above for the catch clauses. *)
+      let catch_state = { state with throw_destination } in
+      let catch_may_throw =
+        catches
+        |> List.fold_left
+             (fun may_throw (_, catch_st) ->
+               let finalcatch, throws =
+                 cfg_stmt_list catch_state (Some catchesi) catch_st
+               in
+               state.g |> add_arc_from_opt (finalcatch, newfakefinally);
+               may_throw || throws)
+             false
+      in
+
+      (* Inside the finally clause, exceptions get propagated. *)
+      let finally_state = state in
+      let finalfinally, finally_may_throw =
+        cfg_stmt_list finally_state (Some newfakefinally) finally_st
+      in
+
+      (* Also propagate any throws that may have happened inside this
+       * try-catch-else-finally statement at the end of the finally clause.
+       * If there's no finally clause, the arcs should already be created
+       * from the code above that sets proper throw_destinations before visiting.
+       *)
+      let may_throw =
+        try_may_throw || else_may_throw || catch_may_throw || finally_may_throw
+      in
+      if may_throw then
```

Comment:
Should this be `may_throw && not has_finally` ? Based on

> If there's no finally clause, the arcs should already be created from the code above that sets proper throw_destinations before visiting.

Reply:
I think the comment added more confusion than clarity, so I'm removing it.

We definitely want to enter this condition when `has_finally` is true. The comment meant to say something like "don't worry, we're not missing arcs that do the propagation".